### DISCUSSION TITLE
Update to KC26 , problems with setting oracle db version

### DISCUSSION THREAD

**Discussion by plantundertable** [keycloak/keycloak#35279] [Q&A] [answered]
I'm trying to update Keycloak to version 26.
Doing so gives me the following Message 
```
ERROR: Persistence unit 'keycloak-default' was configured to run with a database version of at least '19.0.0' but the actual version is '18.0.0' .
```
It then suggests to set 'quarkus.datasource.db-version=18.0.0' in my quarkus.properties.
However the error persists even after doing so. 

After some further investigation and attaching the debugger I noticed that the config is loaded (the set value to 18.0.0 in the recordedConfig Object), but only the version that was set at build time is used (it defaults to 19.0.0).
Below is the class that sets the version that is eventually used.
``` 
package io.quarkus.hibernate.orm.runtime.service;

import java.util.Map;
import java.util.Optional;

import org.hibernate.boot.registry.StandardServiceInitiator;
import org.hibernate.dialect.DatabaseVersion;
import org.hibernate.dialect.Dialect;
import org.hibernate.engine.jdbc.dialect.spi.DialectFactory;
import org.hibernate.service.spi.ServiceRegistryImplementor;

import io.quarkus.hibernate.orm.runtime.recording.RecordedConfig;

public class QuarkusRuntimeInitDialectFactoryInitiator implements StandardServiceInitiator<DialectFactory> {

    private final String persistenceUnitName;
    private final Dialect dialect;
    private final Optional<String> datasourceName;
    private final DatabaseVersion buildTimeDbVersion;

    public QuarkusRuntimeInitDialectFactoryInitiator(String persistenceUnitName, Dialect dialect,
            RecordedConfig recordedConfig) {
        this.persistenceUnitName = persistenceUnitName;
        this.dialect = dialect;
        this.datasourceName = recordedConfig.getDataSource();
        // We set the version from the dialect since if it wasn't provided explicitly through the `recordedConfig.getDbVersion()`
        // then the version from `DialectVersions.Defaults` w
        // ill be used:
        this.buildTimeDbVersion = dialect.getVersion();

    }

    @Override
    public Class<DialectFactory> getServiceInitiated() {
        return DialectFactory.class;
    }

    @Override
    public DialectFactory initiateService(Map<String, Object> configurationValues, ServiceRegistryImplementor registry) {
        return new QuarkusRuntimeInitDialectFactory(persistenceUnitName, dialect, datasourceName, buildTimeDbVersion);
    }
}
```

I experimented with the different Env-Variables of Quarkus and KC but couldn't get it work.

Is it somehow possible to set the version for quarkus at build time with the `kc.sh build ... ` command?
That way I could make sure the correct build time version is set.
Or is there perhaps and even more elegant way to do this by passing some config when running kc with `kc.sh start`

Thank you for reading and if you need further details please let me know.

**Comment by KalvisUpitis** [comment]
I have the same issue with Postgres.
Have tried to set `quarkus.datasource.db-version` and run `kc.sh build` afterwards but with no effect.
Really interesed in the solution.

**Comment by shawkins** [comment, marked as answer]
> I experimented with the different Env-Variables of Quarkus and KC but couldn't get it work.

Unfortunately the underlying issue is this: https://github.com/quarkusio/quarkus/issues/43216

Until we move away from having a persistence.xml in the distribution, you can't properly override the database version.

The only suggested workaround on that issue is to modify the persistence.xml - you should be able to do that by putting an edited version of that in the appropriate package in a jar in the provider directory.

**Reply by plantundertable** [reply]
Thank you for taking the time to give this explanation

**Reply by slominskir** [reply]
I'm running into this as well and wondering what version of persistence.xml to start with and edit.  Where do I find the original?

Poking around I can't find a `persistence.xml`, but I do see there is a `default-persistence.xml` in `libs/org.keycloak-model-jpa-26.0.7.jar`.  The Keycloak source code also references `default-persistence.xml`  at https://github.com/keycloak/keycloak/blob/18afeb83b6687916390af6754d1ef155425cab19/quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java#L403
and
https://github.com/keycloak/keycloak/blob/18afeb83b6687916390af6754d1ef155425cab19/model/jpa/src/main/java/org/keycloak/connections/jpa/util/JpaUtils.java#L63

It's not clear what value I need to add / update, or how to package it back up in the providers directory.   The properties may depend on Hibernate version but could be something like:

```
            <property name="hibernate.dialect" value="org.hibernate.dialect.Oracle12cDialect"/>
```

**Comment by shawkins** [comment]
> Where do I find the original?

Yes the file is called default-persistence.xml - https://raw.githubusercontent.com/keycloak/keycloak/refs/heads/main/model/jpa/src/main/resources/default-persistence.xml

> It's not clear what value I need to add / update

The key should be jakarta.persistence.database-product-version

> or how to package it back up in the providers directory

Put the updated file in at the root of a jar and put it in the providers diretory - it should be seen shadow the built-in one that way.

If we stilll don't get around to pulling out this file, we may have to resort to supporting expressions in it, like shown here https://github.com/quarkusio/quarkus/pull/43240/files#diff-7934b0f96e42a3e76067eaddce2547cf00b4550d1f24e572a9510d017d76d480
